### PR TITLE
[Text Generation] Causal Mask Support

### PR DESCRIPTION
**Pull request by dbogunowicz**
Allows the user to set the argument `prompt_processing_sequence_length` of the `TextGeneration` pipeline to a value different from `sequence_length`, effectively enabling running a `multitoken_engine` in a scenario, when we feed it `input_ids` of any length, and robustly provide kv cache support. In other words, it enables prefilling the cache using subsequences of different lengths.

### Manual Testing

Complementary feature (and PR) from Sparseml: https://github.com/neuralmagic/sparseml/pull/1676

```python
from deepsparse import Pipeline

def _test_pipeline(engine_type):
    opt = Pipeline.create(task="opt",
                          model_path="/home/ubuntu/damian/sparseml/deployment",
                          engine_type=engine_type,
                          prompt_processing_sequence_length=64,
                          use_deepsparse_cache = False,
                          max_generated_tokens=32)
    print('----------')
    prompt = "def hello_world():" # the prompt is short, will not be processed by self.multitoken_engine
    out = opt(sequences=prompt, return_logits = True)
    print(out.sequences[0])
    print('---------')
    prompt = "def hello_world():" * 20 # the prompt is long, will be processed by self.multitoken_engine
    out = opt(sequences=prompt, return_logits=True)
    print(out.sequences[0])

_test_pipeline(engine_type ="onnxruntime")
_test_pipeline(engine_type ="deepsparse")
```

```bash
#### INFERENCE WITH ONNXRUNTIME ####
2023-07-20 05:56:04 deepsparse.transformers.engines.nl_decoder_engine INFO     Overwriting in-place the input shapes of the transformer model at /home/ubuntu/damian/sparseml/deployment/model.onnx
2023-07-20 05:56:10 deepsparse.utils.onnx INFO     Overwriting in-place the batch size of the model at /home/ubuntu/damian/sparseml/deployment/model.onnx
2023-07-20 05:56:18 deepsparse.transformers.engines.nl_decoder_engine INFO     Overwriting in-place the input shapes of the transformer model at /home/ubuntu/damian/sparseml/deployment/model.onnx
2023-07-20 05:56:24 deepsparse.utils.onnx INFO     Overwriting in-place the batch size of the model at /home/ubuntu/damian/sparseml/deployment/model.onnx
----------

    return 'Hello World!'

def hello_world_2():
    return 'Hello World!'

def hello_world_3(): # output is identical to the one from the torch model (baseline)
---------
def hello_world():def hello_world():def hello_world():def hello_world():def hello_world():def hello_world():def hello # output is identical to the one from the torch model (baseline)

#### INFERENCE WITH ONNXRUNTIME ####
# Note that we do not see a warning stating that multitoken engine must default to onnxruntime, because the deepsparse runtime cannot support it! Now both single token and multi token engines are running using deepsparse!

/home/ubuntu/damian/deepsparse/src/deepsparse/transformers/pipelines/text_generation.py:126: UserWarning: AVX512 support not detected, disabling internal management of KV cache which may affect performance. To enable full performance, deploy on an AVX512-compatible system.
  warnings.warn(
Using pad_token, but it is not set yet.
2023-07-20 05:56:37 deepsparse.transformers.engines.nl_decoder_engine INFO     Overwriting in-place the input shapes of the transformer model at /home/ubuntu/damian/sparseml/deployment/model.onnx
DeepSparse, Copyright 2021-present / Neuralmagic, Inc. version: 1.6.0.20230711 COMMUNITY | (34a5203e) (release) (optimized) (system=avx2, binary=avx2)
2023-07-20 05:58:58 deepsparse.transformers.engines.nl_decoder_engine INFO     Overwriting in-place the input shapes of the transformer model at /home/ubuntu/damian/sparseml/deployment/model.onnx
----------

    return 'Hello World!'

def hello_world_2():
    return 'Hello World!'

def hello_world_3():
---------
def hello_world():def hello_world():def hello_world():def hello_world():def hello_world():def hello_world():def hello
```

(in this scenario we run with the default `prompt_processing_sequence_length=64`, but setting it to `32` gives the same result naturally)

### CODE REVIEW

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 1]
```diff
@@ -261,15 +255,33 @@ def process_inputs(self, inputs: TextGenerationInput) -> List[numpy.ndarray]:
...
-        # let's make it optional in the future
         positions = attention_mask.cumsum(1) * attention_mask
         positions -= 1  # assert that positions start at 0
         positions_input = dict(positions=positions)
 
-        input_tokens = {**input_tokens, **positions_input}
+        causal_mask = create_causal_mask(
+            input_tokens["input_ids"], input_tokens["attention_mask"]
+        )
+        causal_mask_input = dict(causal_mask=causal_mask)
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:282]: nit: why do we need to wrap it into a dict, can't we just below do
```python
input_tokens = dict(**input_tokens, positions=positions, causal_mask=causal_mask)
```

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 2]
```diff
@@ -429,12 +432,57 @@ def autoregressive_inference(
...
+        Takes a list of tokens and turns the first
+        `self.prompt_processing_sequence_length` tokens into
+        appropriate engine inputs for the multitoken engine.
+
+        :param tokens: the list of tokens to process
+        :return: a generator of engine inputs
+        """
+        # TODO: Make it yield multiple engine_inputs
+        # Once this is done, we can update the docstrings and be
+        # more verbose about what this function does
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:455]: can you add a few more in-line comments to this function especially the logic branches?
potentially expand the docstring as well?

Also is this function to support multitoken or causal mask? checking understanding as this is the causal mask PR (good either way)

**Review by bfineran — CHANGES_REQUESTED** [review]

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 1 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:282]: yes, what you propose is the same thing but shorter

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 3]
```diff
@@ -261,15 +255,33 @@ def process_inputs(self, inputs: TextGenerationInput) -> List[numpy.ndarray]:
             truncation=truncate,
         )
 
+        if (
+            len(input_tokens[0]) < self.sequence_length
+            and not inputs.fixed_sequences_length
+        ):
+            # if the input is shorter than the sequence length,
+            # we need to pad it to the sequence length
+            padding = "max_length"
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:264]: this second conditional is rubbish, I just realized that couple of days ago but forgot to update :)
